### ISSUE TITLE
[Feature] Implementing `MPI_IN_PLACE` for reductions

### ISSUE THREAD

**Issue by PhilipVinc** [JuliaParallel/MPI.jl#228]
Hi,

I would like to implement functionality to perform reduction with `MPI_IN_PLACE`, so that the output buffer is the same as the input buffer. 
I would gladly contribute this back to the repository, if someone could guide me towards implementing this _correctly_.

For Allreduce I'm thinking about adding a method to `mpi-base.jl` with this signature
```
Allreduce!(sendbuf::MPIBuffertype{T}, op::Union{Op,Function}, comm::Comm) where T = 
    Allreduce!(Ptr{Cvoid}(1), sendbuf, length(sendbuf), op, comm)
```
something similar can be done for `Reduce`. 

Also, `Ptr{Cvoid}(1)` corresponds to the definition of `MPI_IN_PLACE=(void*)1` taken from `mpi.h`. 
I am uncertain on what is the correct way to include it so that it becomes available as `MPI.MPI_IN_PLACE`.

**Comment by PhilipVinc** [comment]
I implemented the above into `philipvinc/in_place` branch at https://github.com/PhilipVinc/MPI.jl/tree/philipvinc/in_place 

At the moment I have added support for 

- Allreduce, Allreducev
- Allgather

If a maintainer confirms that you are interested in this, and helps me with a list of methods to support, I would gladly add those, along with tests, and submit a PR.

**Comment by PhilipVinc** [comment]
Closed by [#232](https://github.com/JuliaParallel/MPI.jl/pull/232)
